src/AppBundle/Resources/config/services.yml

```diff
@@ -23,17 +23,19 @@ services:
         tags:
             - { name: ezpublish.query_type }
 
-    app.criteria.menu:
-        class: AppBundle\Criteria\Menu
+    app.query_type.menu:
+        class: AppBundle\QueryType\MenuQueryType
+        calls:
+            - [setLanguages, [$languages$]]
+        tags:
+            - { name: ezpublish.query_type }
 
     app.controller.menu:
         class: AppBundle\Controller\MenuController
         arguments:
             - '@templating'
             - '@ezpublish.api.service.search'
-            - '@ezpublish.api.service.location'
-            - '@ezpublish.config.resolver'
-            - '@app.criteria.menu'
+            - '@ezpublish.query_type.registry'
```

Comment:
You could simplify dependencies here by passing the QueryType directly: `- @app.query_type.menu`. Then you don't need to use the registry anymore.